Lib/functools.py

```diff
@@ -192,6 +192,212 @@ def total_ordering(cls):
+        nodeinfo.npredecessors += len(predecessors)
+
+        # Create the predecessor -> node edges
+        for pred in predecessors:
+            pred_info = self._get_nodeinfo(pred)
+            pred_info.successors.append(node)
+
+    def prepare(self):
+        """Mark the graph as finished and check for cycles in the graph.
+
+        If any cycle is detected, "CycleError" will be raised, but "get_ready" can
+        still be used to obtain as many nodes as possible until cycles block more
+        progress. After a call to this function, the graph cannot be modified and
+        therefore no more nodes can be added using "add".
+        """
+        if self.ready_nodes is not None:
+            raise ValueError("cannot prepare() more than once")
+
+        self.ready_nodes = [i.node for i in self.node2info.values()
+                            if i.npredecessors == 0]
+        # ready_nodes is set before we look for cycles on purpose:
+        # if the user wants to catch the CycleError, that's fine,
+        # they can continue using the instance to grab as many
+        # nodes as possible before cycles block more progress
+        cycle = self._find_cycle()
+        if cycle:
+            raise CycleError(f"nodes are in a cycle", cycle)
+
+    def get_ready(self):
+        """Return a tuple of all the nodes that are ready.
+
+        Initially it returns all nodes with no predecessors and once those are marked
+        as processed by calling "done", further calls will return all new nodes that
+        have all their predecessors already processed until no more progress can be
+        made.
+
+        Raises ValueError if called without calling "prepare" previously.
+        """
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+
+        # Get the nodes that are ready and mark them
+        result = tuple(self.ready_nodes)
+        n2i = self.node2info
+        for node in result:
+            n2i[node].npredecessors = _NODE_OUT
+
+        # Clean the list of nodes that are ready and update
+        # the counter of nodes that we have returned.
+        self.ready_nodes.clear()
+        self.npassedout += len(result)
+
+        return result
+
+    def is_active(self):
+        """Return True if more progress can be made and ``False`` otherwise.
+
+        Progress can be made if cycles do not block the resolution and either there
+        are still nodes ready that haven't yet been returned by "get_ready" or the
+        number of nodes marked "done" is less than the number that have been returned
+        by "get_ready".
+
+        Raises ValueError if called without calling "prepare" previously.
+        """
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+        return self.nfinished < self.npassedout or bool(self.ready_nodes)
+
+    def __bool__(self):
+        return self.is_active()
+
+    def done(self, node):
```

Comment:
As Tim mentioned on bpo, I wrote one of my own of these recently, and the API all looks very similar.  One minor difference: my done is defined as `def done(self, *nodes):` allowing callers to pass in multiple nodes at once.  Not a big deal but might be a nice minor quality-of-life improvement.

Reply:
Added it in 1efe994!